gateleen-core/src/main/java/org/swisspush/gateleen/core/configuration/ConfigurationResourceManager.java

```diff
@@ -164,63 +177,57 @@ public boolean handleConfigurationResource(final HttpServerRequest request) {
+        String resourceSchema = getRegisteredResources().get(resourceUri);
+        storage.get(resourceUri, buffer -> {
             if (buffer != null) {
                 configurationResourceValidator.validateConfigurationResource(buffer, resourceSchema, event -> {
                     if (event.succeeded()) {
                         if (event.result().isSuccess()) {
-                            notifyObserversAboutResourceChange(requestUri, buffer.toString());
+                            future.complete(Optional.of(buffer));
                         } else {
-                            log.warn("Failure during validation of resource " + requestUri + ". Message: " + event.result().getMessage());
+                            future.fail("Failure during validation of resource " + resourceUri + ". Message: " + event.result().getMessage());
                         }
                     } else {
-                        log.warn("Failure during validation of resource " + requestUri + ". Message: " + event.cause());
+                        future.fail("Failure during validation of resource " + resourceUri + ". Message: " + event.cause());
                     }
                 });
             } else {
-                log.warn("Could not get URL '" + (requestUri == null ? "<null>" : requestUri) + "'.");
+                future.complete(Optional.empty());
             }
         });
+
+        return future;
+    }
+
+    private void notifyObserversAboutRemovedResource(String requestUri) {
+        log.info("About to notify observers that resource " + requestUri + " has been removed");
+        List<ConfigurationResourceObserver> observersByResourceUri = getObserversByResourceUri(requestUri);
+        for (ConfigurationResourceObserver observer : observersByResourceUri) {
+            observer.resourceRemoved(requestUri);
+        }
     }
 
     private void notifyObserverAboutResourceChange(String requestUri, ConfigurationResourceObserver observer) {
-        String resourceSchema = getRegisteredResources().get(requestUri);
-        storage.get(requestUri, buffer -> {
-            if (buffer != null) {
-                configurationResourceValidator.validateConfigurationResource(buffer, resourceSchema, event -> {
-                    if (event.succeeded()) {
-                        if (event.result().isSuccess()) {
-                            observer.resourceChanged(requestUri, buffer.toString());
-                        } else {
-                            log.warn("Failure during validation of resource " + requestUri + ". Message: " + event.result().getMessage());
-                        }
-                    } else {
-                        log.warn("Failure during validation of resource " + requestUri + ". Message: " + event.cause());
+        getValidatedRegisteredResource(requestUri).setHandler(event -> {
+            if(event.failed()){
+                log.warn(event.cause().getMessage());
```

Comment:
The failureMessage already includes the cause (where present). See `future.fail("Failure during validation of resource " + resourceUri + ". Message: " + event.result().getMessage());` and `future.fail("Failure during validation of resource " + resourceUri + ". Message: " + event.cause());`

Reply:
Therefore we're using _String_ to communicate errors rather than _Exception_. Not my taste about how to communicate errors. But that's only my opinion ;)

Reply:
In fact, this could have been implemented "better". On the other side, possible errors are logged, which is what we want here.

However, I think we should focus such code reviews on the more important / mission critical code (and functionality) instead of minor "code styling" issues.